src/ephy-sync-service.c

```diff
@@ -0,0 +1,1417 @@
+ */
+
+#include "config.h"
+#include "ephy-sync-service.h"
+
+#include "ephy-bookmark.h"
+#include "ephy-bookmarks-manager.h"
+#include "ephy-debug.h"
+#include "ephy-fx-password-notification.h"
+#include "ephy-settings.h"
+#include "ephy-shell.h"
+#include "ephy-sync-crypto.h"
+#include "ephy-sync-secret.h"
+
+#include <json-glib/json-glib.h>
+#include <string.h>
+
+#define MOZILLA_TOKEN_SERVER_URL  "https://token.services.mozilla.com/1.0/sync/1.5"
+#define MOZILLA_FXA_SERVER_URL    "https://api.accounts.firefox.com/v1/"
+#define EPHY_BOOKMARKS_COLLECTION "ephy-bookmarks"
+#define EMAIL_REGEX               "^[a-zA-Z0-9_]([a-zA-Z0-9._]+[a-zA-Z0-9_])?@[a-z0-9.-]+$"
+#define SYNC_FREQUENCY            (15 * 60)
+
+struct _EphySyncService {
+  GObject      parent_instance;
+
+  SoupSession *session;
+  guint        source_id;
+
+  char        *uid;
+  char        *sessionToken;
+  char        *keyFetchToken;
+  char        *unwrapBKey;
+  char        *kA;
+  char        *kB;
+
+  char        *user_email;
+  double       sync_time;
+  gint64       auth_at;
+
+  gboolean     locked;
+  char        *storage_endpoint;
+  char        *storage_credentials_id;
+  char        *storage_credentials_key;
+  gint64       storage_credentials_expiry_time;
+  GQueue      *storage_queue;
+
+  char                     *certificate;
+  EphySyncCryptoRSAKeyPair *keypair;
+};
+
+G_DEFINE_TYPE (EphySyncService, ephy_sync_service, G_TYPE_OBJECT);
+
+typedef struct {
+  EphySyncService     *service;
+  char                *endpoint;
+  const char          *method;
+  char                *request_body;
+  double               modified_since;
+  double               unmodified_since;
+  SoupSessionCallback  callback;
+  gpointer             user_data;
+} StorageServerRequestAsyncData;
+
+static StorageServerRequestAsyncData *
+storage_server_request_async_data_new (EphySyncService     *service,
+                                       char                *endpoint,
+                                       const char          *method,
+                                       char                *request_body,
+                                       double               modified_since,
+                                       double               unmodified_since,
+                                       SoupSessionCallback  callback,
+                                       gpointer             user_data)
+{
+  StorageServerRequestAsyncData *data;
+
+  data = g_slice_new (StorageServerRequestAsyncData);
+  data->service = g_object_ref (service);
+  data->endpoint = g_strdup (endpoint);
+  data->method = method;
+  data->request_body = g_strdup (request_body);
+  data->modified_since = modified_since;
+  data->unmodified_since = unmodified_since;
+  data->callback = callback;
+  data->user_data = user_data;
+
+  return data;
+}
+
+static void
+storage_server_request_async_data_free (StorageServerRequestAsyncData *data)
+{
+  if (G_UNLIKELY (!data))
```

Comment:
This doesn't look like a good idea. What's the point of returning early, except to hide bugs? Isn't it better to crash if it's NULL?